### Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio test. The function should return the F-value and the corresponding p-value for the test.

In [3]:
import scipy.stats as stats
import numpy as np

def variance_ratio_test(x, y):
    """
    Function to perform variance ratio test on two arrays of data.
    
    Parameters:
        x (array): First array of data.
        y (array): Second array of data.
    
    Returns:
        F_value (float): F-value for the variance ratio test.
        p_value (float): Corresponding p-value for the test.
    """
    
    # Calculate the variance of x and y
    var_x = np.var(x, ddof=1)
    var_y = np.var(y, ddof=1)
    
    # Calculate the F-value
    F_value = var_x / var_y
    
    # Calculate the degrees of freedom
    df_x = len(x) - 1
    df_y = len(y) - 1
    
    # Calculate the p-value
    p_value = stats.f.sf(F_value, df_x, df_y)
    
    return F_value, p_value


In [4]:
x = [1, 2, 3, 4, 5]
y = [6, 7, 8, 9, 10]

F, p = variance_ratio_test(x, y)

print("F-value: ", F)
print("p-value: ", p)


F-value:  1.0
p-value:  0.5


### Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [5]:
import scipy.stats as stats

def critical_f_value(df_n, df_d, alpha=0.05):
    """
    Function to calculate the critical F-value for a two-tailed test.
    
    Parameters:
        df_n (int): Degrees of freedom for the numerator.
        df_d (int): Degrees of freedom for the denominator.
        alpha (float): Significance level (default is 0.05).
    
    Returns:
        crit_f_value (float): Critical F-value for the two-tailed test.
    """
  
    crit_f_value = stats.f.ppf(1-alpha/2, df_n, df_d)
    
    return crit_f_value


In [6]:
df_n = 3
df_d = 20
alpha = 0.05

crit_f = critical_f_value(df_n, df_d, alpha)

print("Critical F-value: ", crit_f)


Critical F-value:  3.8586986662732143


### Q3. Write a Python program that generates random samples from two normal distributions with known variances and uses an F-test to determine if the variances are equal. The program should output the F- value, degrees of freedom, and p-value for the test.

In [7]:
import numpy as np
import scipy.stats as stats

# Set the random seed for reproducibility
np.random.seed(123)

# Generate random samples from two normal distributions
# with known variances
n1 = 20
n2 = 25
mean1 = 10
mean2 = 12
var1 = 4
var2 = 9

sample1 = np.random.normal(mean1, np.sqrt(var1), n1)
sample2 = np.random.normal(mean2, np.sqrt(var2), n2)

# Calculate the F-value and p-value for the F-test
F_value = np.var(sample1, ddof=1) / np.var(sample2, ddof=1)
df1 = n1 - 1
df2 = n2 - 1
p_value = stats.f.sf(F_value, df1, df2) * 2

# Print the results
print("F-value:", F_value)
print("Degrees of freedom:", df1, "and", df2)
print("p-value:", p_value)


F-value: 0.6419886263062515
Degrees of freedom: 19 and 24
p-value: 1.6725718800022262


### Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from each population. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

To conduct an F-test at the 5% significance level to determine if the variances of two populations are significantly different, we can use the following steps:

1. Set the null and alternative hypotheses:

    * H0: σ1^2 = σ2^2 (the variances are equal)
    * Ha: σ1^2 ≠ σ2^2 (the variances are significantly different)
2. Calculate the F-statistic using the formula:

    F = s1^2 / s2^2

where s1^2 and s2^2 are the sample variances of the two populations.

3. Calculate the degrees of freedom for the F-distribution:

    df1 = n1 - 1

    df2 = n2 - 1

    where n1 and n2 are the sample sizes of the two populations.

4. Calculate the p-value using the F-distribution and the F-statistic:

    p-value = P(F > F_crit) + P(F < -F_crit)
    
where F_crit is the critical F-value for the given degrees of freedom and significance level.

5. Compare the p-value to the significance level. If the p-value is less than the significance level, reject the null hypothesis. Otherwise, fail to reject the null hypothesis.

In [8]:
import scipy.stats as stats

# Set the significance level
alpha = 0.05

# Set the known variances and sample sizes
var1 = 10
var2 = 15
n1 = 12
n2 = 12

# Calculate the F-value
F_value = var1 / var2

# Calculate the degrees of freedom
df1 = n1 - 1
df2 = n2 - 1

# Calculate the critical F-value
F_crit = stats.f.ppf(1-alpha/2, df1, df2)

# Calculate the p-value
p_value = stats.f.sf(F_value, df1, df2) * 2

# Print the results
print("F-value:", F_value)
print("Degrees of freedom:", df1, "and", df2)
print("Critical F-value:", F_crit)
print("p-value:", p_value)

# Compare p-value to significance level
if p_value < alpha:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis")


F-value: 0.6666666666666666
Degrees of freedom: 11 and 11
Critical F-value: 3.473699051085809
p-value: 1.4876102012642005
Fail to reject null hypothesis


### Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance level to determine if the claim is justified.

Set the null and alternative hypotheses:

* H0: σ^2 = 0.005 (the variance is equal to the claimed value)
* Ha: σ^2 ≠ 0.005 (the variance is significantly different from the claimed value)

In [12]:
import scipy.stats as stats

# Set the significance level
alpha = 0.01

# Set the claimed variance, sample variance, and sample size
var_claimed = 0.005
var_sample = 0.006
n = 25

# Calculate the F-value
F_value = var_sample / var_claimed

# Calculate the degrees of freedom
df1 = n - 1
df2 = n - 1

# Calculate the critical F-value
F_crit = stats.f.ppf(1-alpha/2, df1, df2)

# Calculate the p-value
p_value = stats.f.sf(F_value, df1, df2) * 2

# Print the results
print("F-value:", F_value)
print("Degrees of freedom:", df1, "and", df2)
print("Critical F-value:", F_crit)
print("p-value:", p_value)

# Compare p-value to significance level
if p_value < alpha:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis")


F-value: 1.2
Degrees of freedom: 24 and 24
Critical F-value: 2.966741631292762
p-value: 0.6587309365634489
Fail to reject null hypothesis


### Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an F-distribution and calculates the mean and variance of the distribution. The function should return the mean and variance as a tuple.

In [13]:
import numpy as np

def f_distribution_stats(df1, df2):
    """
    Calculate the mean and variance of an F-distribution given the degrees
    of freedom for the numerator and denominator.
    
    Parameters:
        df1 (int): Degrees of freedom for the numerator.
        df2 (int): Degrees of freedom for the denominator.
    
    Returns:
        A tuple containing the mean and variance of the F-distribution.
    """
    if df1 <= 0 or df2 <= 0:
        raise ValueError("Degrees of freedom must be positive.")
    
    mean = df2 / (df2 - 2)
    variance = (2 * df2**2 * (df1 + df2 - 2)) / (df1 * (df2 - 2)**2 * (df2 - 4))
    
    return mean, variance


In [16]:
mean, variance = f_distribution_stats(5,6)
print(f"mean :{mean}\nvariance : {variance}")

mean :1.5
variance : 4.05


### Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The sample variance is found to be 25. Another random sample of 15 measurements is taken from another normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test at the 10% significance level to determine if the variances are significantly different.


To determine if the variances of the two populations are significantly different, we can conduct an F-test at the 10% significance level.
The null hypothesis is that the variances are equal, 
and the alternative hypothesis is that they are not equal.

In [18]:
import scipy.stats as stats

# sample sizes and variances
n1 = 10
s1_squared = 25
n2 = 15
s2_squared = 20

# calculate the F-test statistic
F = s1_squared / s2_squared

# degrees of freedom
df1 = n1 - 1
df2 = n2 - 1

# calculate the p-value
p_value = 2 * (1 - stats.f.cdf(F, df1, df2))

# calculate the critical F-value at the 10% significance level
alpha = 0.1
f_critical = stats.f.ppf(alpha/2, df1, df2)

# print the results
print("F-test statistic: ", F)
print("Degrees of freedom: ", df1, " and ", df2)
print("Critical F-value: ", f_critical)
print("p-value: ", p_value)

if F > f_critical:
    print("Reject the null hypothesis: the variances are significantly different.")
else:
    print("Fail to reject the null hypothesis: the variances are not significantly different.")


F-test statistic:  1.25
Degrees of freedom:  9  and  14
Critical F-value:  0.3305268601412525
p-value:  0.6832194382585954
Reject the null hypothesis: the variances are significantly different.


The F-test statistic is 1.25, the degrees of freedom are 9 and 14, and the critical F-value at the 10% significance level is 0.3305. The p-value is approximately 0.6832 ,and is less than 0.1 we fail to reject the null hypothesis that the variances are equal

### Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

In [19]:
import numpy as np
from scipy.stats import f

# Sample data for Restaurant A and B
a = np.array([24, 25, 28, 23, 22, 20, 27])
b = np.array([31, 33, 35, 30, 32, 36])

# Sample variances
var_a = np.var(a, ddof=1)
var_b = np.var(b, ddof=1)

# F-test
f_value = var_a / var_b
dfn = len(a) - 1
dfd = len(b) - 1
p_value = 2 * min(f.cdf(f_value, dfn, dfd), 1 - f.cdf(f_value, dfn, dfd))

# Output results
print("Sample variance of Restaurant A:", var_a)
print("Sample variance of Restaurant B:", var_b)
print("F-test statistic:", f_value)
print("Degrees of freedom:", dfn, "and", dfd)
print("Critical F-value:", f.ppf(0.025, dfn, dfd), "and", f.ppf(0.975, dfn, dfd))
print("p-value:", p_value)


Sample variance of Restaurant A: 7.80952380952381
Sample variance of Restaurant B: 5.366666666666667
F-test statistic: 1.4551907719609583
Degrees of freedom: 6 and 5
Critical F-value: 0.16701279718024772 and 6.977701858535566
p-value: 0.6974815747937484


Since the p-value of 0.34 is greater than the significance level of 0.05, we cannot reject the null hypothesis that the variances are equal. Therefore, we do not have sufficient evidence to conclude that the variances of waiting times at the two restaurants are significantly different.

### Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83; Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances are significantly different.

In [20]:
import numpy as np
from scipy import stats

# Group A data
group_a = np.array([80, 85, 90, 92, 87, 83])
var_a = np.var(group_a, ddof=1)  # sample variance of Group A

# Group B data
group_b = np.array([75, 78, 82, 79, 81, 84])
var_b = np.var(group_b, ddof=1)  # sample variance of Group B

# Test statistic F
f = var_a / var_b if var_a > var_b else var_b / var_a
print("F-test statistic: ", round(f, 2))

# Degrees of freedom
df_num = len(group_a) - 1  # degrees of freedom for numerator
df_den = len(group_b) - 1  # degrees of freedom for denominator
print("Degrees of freedom: ", df_num, " and ", df_den)

# Critical F-value
alpha = 0.01  # significance level
crit_f = stats.f.ppf(alpha/2, df_num, df_den)
print("Critical F-value: ", round(crit_f, 2))

# p-value
p_val = stats.f.sf(f, df_num, df_den) * 2
print("p-value: ", round(p_val, 2))

# Conclusion
if f > crit_f:
    print("Reject null hypothesis, variances are significantly different.")
else:
    print("Fail to reject null hypothesis, variances are not significantly different.")


F-test statistic:  1.94
Degrees of freedom:  5  and  5
Critical F-value:  0.07
p-value:  0.48
Reject null hypothesis, variances are significantly different.


The calculated F-value is 1.94 and the critical F-value at a 1% significance level with 5 and 5 degrees of freedom is 0.07. Since the calculated F-value is greater than the critical F-value, we reject the null hypothesis that the variances of the two groups are equal. Therefore, we conclude that the variances are significantly different at the 1% significance level.